bin/check-megaraid-sas-status.rb

```diff
@@ -58,7 +58,7 @@ def run
     (0..$CHILD_STATUS.exitstatus - 1).each do |i|
       # and check them in turn
       stdout = `#{config[:megaraidcmd]} -LDInfo -L#{i} -a#{config[:controller]} `
-      unless Regexp.new('State\s*:\s*Optimal').match?(stdout)
+      unless Regexp.new('State\s*:\s*Optimal').match(stdout)
```

Comment:
I am having a hard time understanding why this does not work since we don't assign the returned match data anyways. From my tests it seems to have no effect.
```ruby
$ irb
# assign vars
irb(main):001:0> a = Regexp.new('State: Optimal')
=> /State: Optimal/
irb(main):002:0> b = 'State: Optimal foo bar'
=> "State: Optimal foo bar"

# test using match?
irb(main):003:0> unless a.match?(b)
irb(main):004:1> msg = sprintf '%svirtual drive %d: %s ', error, i, b[/State\s*:\s*.*/].split(':')[1]
irb(main):005:1> p msg
irb(main):006:1> end
=> nil

# testing using match
irb(main):007:0> unless a.match(b)
irb(main):008:1> msg = sprintf '%svirtual drive %d: %s ', error, i, b[/State\s*:\s*.*/].split(':')[1]
irb(main):009:1> p msg
irb(main):010:1> end
=> nil
```

So I can better understand can you please include the output of adding `p stdout: #{stdout}, class: #{stdout.class}` before and after the `unless`?

Reply:
Does the following output with stdout before the unless suffice?
https://gist.github.com/crscybr/3aca2c9d0e61457581ad035ca91b9dc9

Reply:
It does actually help, I think I understand what is happening now, just to confirm can you verify if you are using a ruby that is `< 2.4.0`? The function `match?` was added in ruby 2.4 so that explains the error. I was under the impression that it was silently passing when it should have failed or failing when it should have passed. As this was an uncaught error I did not look to see if it was not available in certain versions of ruby. In addition to this script there are others that we should fix up. Any chance you would be willing to submit a PR to fix those up as well? If not I can maybe get to it tomorrow.

Reply:
This is why including a before and after run is helpful sometimes just by looking at the actual error I can quickly validate the fix much quicker.

Reply:
Perhaps the version info is relevant.  The test above also fails on my system using sensu-0.26.5-2.x86_64
% /opt/sensu/embedded/bin/ruby -v
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-linux]

```/opt/sensu/embedded/bin/irb
irb(main):001:0> a = Regexp.new('State: Optimal')
=> /State: Optimal/
irb(main):002:0> b = 'State: Optimal foo bar'
=> "State: Optimal foo bar"
irb(main):003:0> unless a.match?(b)
irb(main):004:1> msg = sprintf '%svirtual drive %d: %s ', error, i, b[/State\s*:\s*.*/].split(':')[1]
irb(main):005:1> p msg
irb(main):006:1> end
NoMethodError: undefined method `match?' for /State: Optimal/:Regexp
Did you mean?  match
	from (irb):3
	from /opt/sensu/embedded/bin/irb:11:in `<main>'
irb(main):007:0>
```

Reply:
Yup that confirms it, this borke on ruby versions `< 2.4.0`.

Reply:
Pardon my ignorance but would something like `unless stdout =~ /State\s*:\s*Optimal/`  work more broadly ?

Reply:
Sorry if my replies are erratic, I am on vacation this week.

Correct that should work based on my understanding of what each of the following does:
- `=~ //` returns the index of the index (integer) of the first match in the string otherwise returns `nil` if no match is found. Updates global var `$~`.
- `.match()` returns the a  `MatchData` object which contains the matching data and useful metadata such as capture groups otherwise `nil` if no match is found. Updates global var `$~`.
- `.match?()` which was introduced in `2.4` returns a `(true|false)` and does not update any global vars such as `$~`

Since we don't need anything other than knowing if the data matches in this case we should use `=~` for the time being to keep fix the bug that this affects in ruby `< 2.4` setups. We will probably change that to `match?` at some point when we drop ruby `< 2.4` support as it is clearer in its intent for a ruby novice. Previously `.match()` was used and while we could restore it I think it would be best to fall back to `=~` as this is closer to the `.match?` implementation and therefore is more performant.

Sources:
- https://ruby-doc.org/core-2.4.1/doc/regexp_rdoc.html
- https://ruby-doc.org/core-2.4.1/Regexp.html#method-i-match-3F